### DISCUSSION TITLE
What's the right way to use NuxtErrorBoundary?

### DISCUSSION THREAD

**Discussion by xhc-code** [nuxt/nuxt#20223] [Questions] [answered]
### Discussed in https://github.com/nuxt/nuxt/discussions/16736

<div type='discussions-op-text'>

<sup>Originally posted by **xiangnanscu** September 28, 2022</sup>
I press th Submit button to throw an error in client, but nothing shows up, including the console.
```vue
pages/login.vue
<template>
  <div>
    <NuxtErrorBoundary @error="someErrorLogger">
      <!-- You use the default slot to render your content -->
      <template #error="{ error }">
        You can display the error locally here.
        <button @click="error = null">This will clear the error.</button>
      </template>
    </NuxtErrorBoundary>
    <x-button @click.prevent="sendCredentials">Submit</x-button>
  </div>
</template>

<script setup>
function someErrorLogger(event) {
  console.log('error happend:', { event });
}
async function sendCredentials(event) {
 throw createError({ statusCode: 404, statusMessage: "Page Not Found" });
}
</script>

```
</div>

**Comment by danielroe** [comment]
The thing producing the error must be 'within' (that is, a child of) the error boundary.

**Reply by xhc-code** [reply]
The NuxtErrorBoundary component has no effect, and the official example has no actual usage of this component.
Subcomponents wrapped by the NuxtErrorBoundary component, throwing Error or createError provided by nuxt cannot make the error slot of the NuxtErrorBoundary component display content. The usage of this is questionable.

**Reply by xhc-code** [reply]
> The NuxtErrorBoundary component has no effect, and the official example has no actual usage of this component. Subcomponents wrapped by the NuxtErrorBoundary component, throwing Error or createError provided by nuxt cannot make the error slot of the NuxtErrorBoundary component display content. The usage of this is questionable.

This official example is old, don't see where the NuxtErrorBoundary component is actually referenced.
https://nuxt.com/docs/examples/app/error-handling

**Reply by danielroe** [reply, marked as answer]
Here's an example.

https://stackblitz.com/edit/github-wg3yu6

And here are the docs: https://nuxt.com/docs/api/components/nuxt-error-boundary#nuxterrorboundary. I'd welcome a PR to update things if we need to.

**Reply by xhc-code** [reply]
Thank you for your example. I understand the usage. The subcomponent wrapped by the NuxtErrorBoundary component needs to ensure that the thrown error is at the onmount hook position, otherwise the thrown error will be taken over and displayed by the error.vue global exception page.
This does not seem to indicate which callbacks are limited to in the position of the document, because I am using the vue3 setup combined page here, so I thought it might be useful in the setup.

Thanks again!

**Reply by cuebit** [reply]
@danielroe I might jump on this if no PR comes through in the next few days. I've seen this question asked a lot lately in Discord, it might make sense to be a little layman for those finding it difficult to adopt error strategies. The error handling example currently incorporates almost every angle of error handling which can be overwhelming for those unfamiliar with it.

**Reply by danielroe** [reply]
That would be amazing ❤️

**Reply by awacode21** [reply]
@danielroe i have a question. In the example you posted you use `throw new Error('bob')`. i also see t`hrow new Error` used in many other examples. But in Nuxt documentation its saying we should use `throw CreateError`. So i am a bit confused when to use what?

**Reply by danielroe** [reply]
I would always use createError if it's your code. That gives you some extra options and plays nice with serialising/deserialising errors with asyncData.

**Reply by awacode21** [reply]
@danielroe i have an additional question on how to use NuxtErrorBoundary right. When i use createError for example like this:

```
throw createError({
      statusCode: error.value.statusCode,
      statusMessage: error.value.statusMessage,
      message: 'Die Datenansicht konnte auf Grund von Validierungsproblemen nicht geladen werden.',
    });
```
    
  then i would expect that i can access the message 

> Die Datenansicht konnte auf Grund von Validierungsproblemen nicht geladen werden.

within my NuxtErrorBoundary error slot
  
```
<template>
  <NuxtErrorBoundary>
    <template #error="{ error, clearError }">
      {{ error.message }}
      <button @click="clearError">Clear Error.</button>
    </template>
    <slot></slot>
  </NuxtErrorBoundary>
</template>
```

but it is not working. `error.message` does not exist. 
When i use this instead:


```
<template>
  <NuxtErrorBoundary>
    <template #error="{ error, clearError }">
      {{ error }}
      <button @click="clearError">Clear Error.</button>
    </template>
    <slot></slot>
  </NuxtErrorBoundary>
</template>
```

then i get displayed on screen:

<img width="715" alt="Bildschirmfoto 2023-11-24 um 19 22 48" src="https://github.com/nuxt/nuxt/assets/30939926/52e8addc-b589-4042-ac61-5ec5e9b6bfe3">


**So how can i show a custom error message without prefixing it with Error: and surrounding it with quotation marks?**

**Reply by al1maher** [reply]
Same issue with me. Found a solution?

**Reply by tag** [reply]
{{ error.message }} or {{ error.statusMessage }}

**Comment by awacode21** [comment]
@danielroe second question: when i normally want to handle errors within the NuxtErrorBoundary .. but for specific status codes like a 500 i want the global error.vue page. How can i manage that? cause when i use fatal:true in that case, its still showing my error within the NuxtErrorBoundary and not directing to error.vue

**Reply by awacode21** [reply]
well, this one i get to work using showError instead of createError. It don't know if this is the recommended way, but it would work
